src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsHash.cs

```diff
@@ -0,0 +1,109 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    internal abstract class CmsHash : IDisposable
+    {
+        internal static CmsHash Create(Oid oid, bool forVerification) =>
+            oid.Value switch
+            {
+#if NET
+                Oids.Shake128 => new CmsShake128Hash(),
+                Oids.Shake256 => new CmsShake256Hash(),
+#endif
+                _ => new CmsIncrementalHash(oid, forVerification),
+            };
+
+        public abstract void Dispose();
+        internal abstract void AppendData(ReadOnlySpan<byte> data);
+        internal abstract byte[] GetHashAndReset();
+
+#if NET || NETSTANDARD2_1
+        internal abstract bool TryGetHashAndReset(Span<byte> destination, out int bytesWritten);
+#endif
+
+#if NET
+        private sealed class CmsShake256Hash : CmsHash
+        {
+            // RFC 8702 specifies SHAKE256 in CMS must use 512 bits of output.
+            private const int OutputSizeBytes = 512 / 8;
+
+            private readonly Shake256 _shake256;
```

Comment:
YESSS. Finally some use of SHAKE!

Note that we did not OOB any of the SHA-3 APIs. So, if you think this is going to be needed to make CMS work for .NET Standard or .NET 9, then the Shake API's won't be available.

Reply:
Talked offline, decided to only support .NET.